File: tests/coretests/Stocks/Services/MultipleBarPriceAnalysisTests.cs
using System.Collections.Generic;
using System.Linq;
using core.fs.Services;
using core.fs.Services.Analysis;
using coretests.testdata;
using Xunit;

namespace coretests.Stocks.Services
{

    public static class AnalysisOutcomeExtensions
    {
        public static AnalysisOutcome FirstOutcome(this IEnumerable<AnalysisOutcome> outcomes, string key) => outcomes.First(x => x.Key == key);
    }

    public class MultipleBarPriceAnalysisTests
    {
        private readonly List<AnalysisOutcome> _outcomes =
            MultipleBarPriceAnalysis.MultipleBarPriceAnalysis.run(
                TestDataGenerator.PriceBars(TestDataGenerator.NET)
            );

        [Fact]
        public void OutcomesMatch() => Assert.NotEmpty(_outcomes);

        [Fact]
        public void PercentAboveLow()
            => Assert.Equal(0.3m, _outcomes.FirstOutcome(MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.PercentAboveLow).Value, 2);

        [Fact]
        public void PercentBelowHigh()
            => Assert.Equal(0.77m, _outcomes.FirstOutcome(MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.PercentBelowHigh).Value, 2);

        [Fact]
        public void PercentChangeAverage()
            => Assert.Equal(-0.16m, _outcomes.FirstOutcome(MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.PercentChangeAverage).Value, 2);

        [Fact]
        public void PercentChangeStandardDeviation()
            => Assert.Equal(6.62m, _outcomes.FirstOutcome(MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.PercentChangeStandardDeviation).Value, 2);

        [Fact]
        public void LowestPrice()
            => Assert.Equal(37.84m, _outcomes.FirstOutcome(MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.LowestPrice).Value, 2);

        [Fact]
        public void LowestPriceDaysAgo()
            => Assert.True(
                _outcomes.FirstOutcome(MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.LowestPriceDaysAgo).Value > 345 // it keeps on increasing as time goes by and test data's date is static
            );

        [Fact]
        public void HighestPrice()
            => Assert.Equal(217.25m, _outcomes.FirstOutcome(MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.HighestPrice).Value, 2);

        [Fact]
        public void HighestPriceDaysAgo()
            => Assert.True(
                _outcomes.FirstOutcome(MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.HighestPriceDaysAgo).Value > _outcomes.FirstOutcome(MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.LowestPriceDaysAgo).Value
            );
        
        [Fact]
        public void CurrentPrice()
            => Assert.Equal(49.14m, _outcomes.FirstOutcome(MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.CurrentPrice).Value, 2);
        
        [Fact]
        public void EarliestPrice()
            => Assert.Equal(75.08m, _outcomes.FirstOutcome(MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.EarliestPrice).Value, 2);
        
        [Fact]
        public void Gain()
            => Assert.Equal(-0.35m, _outcomes.FirstOutcome(MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.Gain).Value, 2);
        
    }
}

File: tests/coretests.fs/Stocks/Services/MultipleBarPriceAnalysisTests.fs
module coretests.fs.Stocks.Services.MultipleBarPriceAnalysisTests

open Xunit
open core.fs.Services
open core.fs.Services.Analysis
open testutils
open FsUnit

let outcomesWithSmallNumberOfBars = MultipleBarPriceAnalysis.run (TestDataGenerator.IncreasingPriceBars(10));
let rounded (decimals:int) (value:decimal) = System.Math.Round(value, decimals)

[<Fact>]
let ``Outcomes for small number of bars are not empty``() =
    Assert.NotEmpty(outcomesWithSmallNumberOfBars)

let firstOutcomeGeneric key (outcomes:AnalysisOutcome list) = outcomes |> Seq.find(fun x -> x.Key = key) |> _.Value

let outcomes = MultipleBarPriceAnalysis.run (TestDataGenerator.PriceBars(TestDataGenerator.NET))

let private firstOutcome key = firstOutcomeGeneric key outcomes

[<Fact>]
let ``Outcomes for large number of bars are not empty``() =
    Assert.NotEmpty(outcomes)

[<Fact>]
let ``Percent above low outcome correct``() =
    Assert.Equal(0.3m, firstOutcome MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.PercentAboveLow, 2)

[<Fact>]
let ``Percent below high outcome correct``() =
    Assert.Equal(0.77m, firstOutcome MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.PercentBelowHigh, 2)

[<Fact>]
let ``Percent change average outcome correct``() =
    Assert.Equal(-0.0016m, firstOutcome MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.PercentChangeAverage, 4)

[<Fact>]
let ``Percent change standard deviation outcome correct``() =
    Assert.Equal(0.0662m, firstOutcome MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.PercentChangeStandardDeviation, 4)

[<Fact>]
let ``Lowest price outcome correct``() =
    Assert.Equal(37.84m, firstOutcome MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.LowestPrice, 2)

[<Fact>]
let ``Lowest price days ago outcome correct``() =
    Assert.True(firstOutcome MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.LowestPriceDaysAgo > 345m) // it keeps on increasing as time goes by and test data's date is static

[<Fact>]
let ``Highest price outcome correct``() =
    Assert.Equal(217.25m, firstOutcome MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.HighestPrice, 2)

[<Fact>]
let ``Highest price days ago outcome correct``() =
    let highestDaysAgo = firstOutcome MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.HighestPriceDaysAgo
    let lowestDaysAgo = firstOutcome MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.LowestPriceDaysAgo
    highestDaysAgo > lowestDaysAgo |> should be True

[<Fact>]
let ``Current price outcome correct``() =
    firstOutcome MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.CurrentPrice |> should equal 49.14m
    
[<Fact>]
let ``Earliest price outcome correct``() =
    firstOutcome MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.EarliestPrice |> should equal 75.08m

[<Fact>]
let ``Gain should be correct()`` () =
    firstOutcome MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.Gain |> rounded 2 |> should equal -0.35m

[<Fact>]
let ``Average true range should be correct`` () =
    firstOutcome MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.AverageTrueRange |> rounded 2 |> should equal 3.98m
    
[<Fact>]
let ``Green streak is positive`` () =
    firstOutcome MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.GreenStreak |> should equal 1
    
[<Fact>]
let ``EMA20 above SMA50 is negative and matches`` () =
    firstOutcome MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.EMA20AboveSMA50Bars |> should equal -48m
    
[<Fact>]
let ``Price above EMA20 bars is positive and matches`` () =
    firstOutcome MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.PriceAboveEMA20Bars |> should equal 1m

[<Fact>]
let ``SMA50 above SMA200 bars is positive and matches`` () =
    firstOutcome MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.SMA50AboveSMA200Bars |> should equal -205m
    
[<Fact>]
let ``On Balance Volume is present and matches`` () =
    firstOutcome MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.OnBalanceVolume |> should equal 41298923m
    
[<Fact>]
let ``Average True Range Percentage is present and matches`` () =
    firstOutcome MultipleBarPriceAnalysis.MultipleBarOutcomeKeys.AverageTrueRangePercentage |> rounded 2 |> should equal 0.08m
